Signal/src/call/PeerConnectionClient.swift

```diff
@@ -326,115 +404,145 @@ class PeerConnectionClient: NSObject, RTCPeerConnectionDelegate, RTCDataChannelD
         Logger.debug("\(TAG) dataChannelDidChangeState: \(dataChannel)")
     }
 
     /** The data channel successfully received a data buffer. */
-    public func dataChannel(_ dataChannel: RTCDataChannel, didReceiveMessageWith buffer: RTCDataBuffer) {
-        Logger.debug("\(TAG) dataChannel didReceiveMessageWith buffer:\(buffer)")
-
-        guard let dataChannelMessage = OWSWebRTCProtosData.parse(from:buffer.data) else {
-            // TODO can't proto parsings throw an exception? Is it just being lost in the Objc->Swift?
-            Logger.error("\(TAG) failed to parse dataProto")
-            return
-        }
+    internal func dataChannel(_ dataChannel: RTCDataChannel, didReceiveMessageWith buffer: RTCDataBuffer) {
+        PeerConnectionClient.signalingQueue.async {
+            Logger.debug("\(self.TAG) dataChannel didReceiveMessageWith buffer:\(buffer)")
+
+            guard let dataChannelMessage = OWSWebRTCProtosData.parse(from:buffer.data) else {
+                // TODO can't proto parsings throw an exception? Is it just being lost in the Objc->Swift?
+                Logger.error("\(self.TAG) failed to parse dataProto")
+                return
+            }
 
-        if let delegate = delegate {
-            delegate.peerConnectionClient(self, received: dataChannelMessage)
+            if let delegate = self.delegate {
+                DispatchQueue.main.async {
+                    delegate.peerConnectionClient(self, received: dataChannelMessage)
+                }
+            }
         }
     }
 
     /** The data channel's |bufferedAmount| changed. */
-    public func dataChannel(_ dataChannel: RTCDataChannel, didChangeBufferedAmount amount: UInt64) {
+    internal func dataChannel(_ dataChannel: RTCDataChannel, didChangeBufferedAmount amount: UInt64) {
         Logger.debug("\(TAG) didChangeBufferedAmount: \(amount)")
     }
 
     // MARK: - RTCPeerConnectionDelegate
 
     /** Called when the SignalingState changed. */
-    public func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {
+    internal func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {
         Logger.debug("\(TAG) didChange signalingState:\(stateChanged.debugDescription)")
     }
 
     /** Called when media is received on a new stream from remote peer. */
-    public func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
-        Logger.debug("\(TAG) didAdd stream:\(stream) video tracks: \(stream.videoTracks.count) audio tracks: \(stream.audioTracks.count)")
-
-        if stream.videoTracks.count > 0 {
-            remoteVideoTrack = stream.videoTracks[0]
-            if let delegate = delegate {
-                delegate.peerConnectionClient(self, didUpdateRemote: remoteVideoTrack)
+    internal func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
```

Comment:
Some of these `RTCPeerConnection` delegate callbacks were actually being invoked on a WebRTC thread. That could have caused some hard-to-reproduce bugs.